### DISCUSSION TITLE
Medium and Large models hallucination

### DISCUSSION THREAD

**Discussion by DavraYoung** [openai/whisper#678] [Q&A] [answered]
Hi, I am trying to train Whisper Medium model on my custom dataset. (Using this [guide](https://huggingface.co/blog/fine-tune-whisper))

I did training on small models and it achieves very good results with WER 10 on my custom dataset.
However using this dataset during medium model training seems to enter failure mode.

By hallucination I mean increasing WER with repeating words. For example my Medium model after training on 500 hours of audio gives me:
Truth: faqat basseyn butun qolgan
Prediction: faqat basseyni butun qolgani vaqtlari vaqtlari vaqtlari vaqtlari vaqtlari vaqtlari vaqtlar

word `vaqtlar` is not present in audio, nor in the original text, the model just lagged on it somehow.

I checked audio itself, and its fine, there are some background noises, but the voice is very recognizable, and in fact my finetuned `small` model recognizes it correctly.

The model behaves better with smaller generation_max_length setting, but cannot predict long audios.

The dataset itself contains small audios within 10 seconds range.

I tried:
1. reducing generation_max_length to the smallest minimum possible
2. reducing learning rate and increasing batch size
3. setting generation_num_beams to 3 or 5
4. removing all punctuation and other characters that are non word related

still no good.

How to solve such issue?

**Comment by KaiserChr** [comment]
I just posted about this a minute ago, I had the same problem. 
Maybe try my (possible) fix described in #679 and let me know if it helps!

**Comment by DavraYoung** [comment, marked as answer]
For any one facing such issue in transformers.
1. Check your dataset, if you have too many incorrect clips (with silence or incorrect audio format, sample rate should be exactly 16000)
2. Check your clips length, it should be cut if you have more than 30 seconds, otherwise you may face issues with padding and hallucination
3. Use latest transformers version
4. Reduce learning rate and batch size (in my case big batch size and learning rate(1e-4 and 80batch size with accumulation 2) made generalization worse and sometimes even caused issues with hallucination, not sure why). In my case the model did very well with 1e-5/2 learning rate and batch size 32 on medium model.

In my case, the dataset was fine, and I managed to overcome the problem by simply updating transformers version to 4.25 and finetuning parameters

Previously I used 4.24, which was causing this, after updating it went well. I suppose the issue was in tokenizer, but I am not sure why exactly

P.S. dont forget to clear your dataset cache and rerun feature_extraction and tokenization again. In my case using newest version with old cache did not work either

**Reply by dgoryeo** [reply]
Thanks for this @DavraYoung. Quick question from a newbie:  how do clear the dataset cache and rerun feature_extraction and tokenization? --Thanks!

**Reply by DavraYoung** [reply]
if you dont provide cache_dir, then cache dirrectory is under ~/.cache/huggingface

**Reply by huynhthanh98** [reply]
Hi!
I trained whisper-medium on my custom dataset with 100 hours of audio. The result of my model in 1st epoch was good with 0.24 wer but after 1st epoch my model was overfitting and i am also facing hallucination.

My setting is:

```
learning_rate : 0.5e-5
weight_decay : 0.01
adam_epsilon : 1.0e-8
warmup_steps : 2
batch_size : 4
num_worker : 4
num_train_epochs : 15
```
my dataset was crawled from google api and i chunked each audio to 30s.

Thanks for your helping.

**Reply by a-nahar** [reply]
Hi @DavraYoung I am also facing the similar issue, but with whisper-small, basically finetuning for more steps causes more hallucinations.
I have around 200 hours of audios(audio duration distribution lies in 15 to 30 sec), and I am fine tuning it for the translation task. the batch size I am using is 8 with grad accumulatio of 4. 
I tried all the things which you mentioned but no as such gains.
as my dataset contains audio of duration between 15 to 30 sec, so I set the generation_max_length to 448(max), 
could you please let me know, if there is something else I can try?
I have tested the model with checkpoint-3000 (3000 steps, it only has 1% gain compare to whisper small). rest all the checkpoints are performing worse than the base model.

**Reply by DavraYoung** [reply]
@a-nahar
Check your dataset. 
Check your transformers version.
Check your data collator.

Most of the time, if you observe 90%+ hallucinations, it means:
1. Your dataset is incorrect. Either audio is corrupted or is shifted in time or is in incorrect sample rate.
2. You are masking endoftext token when shifting tokens right or not including it completely, thus preventing the model from learning where to stop
3. You dont wait long enogh, initially, the model will hallucinate during training, but in my case after 5000-10000 steps usually it stops doing it


1. For 1s case, you can simply save your audio in processing function to check if your audio is actually correct
2. For 2, you need to check what `decoder_input_ids` is supplied to forward method during training, if you simply supply labels, i recommend to shift them and supply as `decoder_input_ids` using data collator as well. For proper Data collator implementation you can check: https://github.com/huggingface/distil-whisper/blob/e35cbff09e04f6e710c72f33e0441d6bebc47ec1/training/run_distillation.py#L400

**Reply by DavraYoung** [reply]
oh, btw, you should check what is your inference forced_token_ids and tokenizer prefix tokens. and see if there is any difference from the on in the training.

I found that model tend to hallucinate if the prefix tokens are different.


For example:
during training model see: `<|startoftranscript|><|transcribe|><|en|> ....transcription <|endoftext|>`
during inference you supply only: `<|transcribe|><|en|>...`, the model may hallucinate because of missing `<|startoftranscript|>`

So make sure your inference prompt does fully match the training one

**Reply by a-nahar** [reply]
Thanks @DavraYoung , got your point. Will look at it.

**Reply by a-nahar** [reply]
Hello @DavraYoung,

I've thoroughly reviewed the things you mentioned, and upon cross-checking, I found no discrepancies. even for the inference I am supplying <|startoftranscript|><|translate|><|hi|> ....transcription <|endoftext|> (translation mode) .


Additionally, I experimented with fine-tuning the large-v2 model and found it surprisingly free from hallucinations post-fine-tuning. However, when comparing the results of the fine-tuned model with the pretrained large-v2, it's not conclusively better. Roughly 50% of the testing samples, the fine-tuned model outperforms the pretrained large-v2, while the remaining 50% it underperforms. Through a macro analysis of the underperforming examples, I noticed a pattern – these instances tend to have shorter transcriptions( it could be because the <|endoftext|>token is getting predicted in the middle of the audio).

Any suggestions from your end?

Regarding the large-v2 model, I found the most optimal checkpoint at 4000 steps, utilizing a total of 250+ audio data points. Surprisingly, after this point, the evaluation BLEU (higher scores are better) began to decrease, could be bcoz of the overfitting.

may be hallucination is a tendency of the small and medium model (I may be wrong here).

**Reply by DavraYoung** [reply]
@a-nahar if the model predicts endoftext too early, most probably your dataset contains mostly short audios and the model is overfitted to predict no more than certain number of tokens.

I would suggest:
1. Try to mix your samples with longer samples (e.g. increase the number of long text audios), or mix it with long audios dataset
2. Try to experiment with logits processors. There is a way to change the probability during token generation, theoretically, you can decrease the probability of endoftext token so that it happens less often.

**Reply by Stanwang1210** [reply]
Hi @DavraYoung 
I'm also working on a project to do low-resource fine-tuning whisper with lora.
I also face some hallucination problem recently.
I found that when I fine-tune whisper-medium to German (A seen languages for whisper) with one-hour of data, the ASR performance of German on test set degrades (CER : 5.4% -> 21.7%), which seems weird. 
During the log file, I did find some hallucination issues (some of the samples). I'm wondering whether there is bug in my code. 
The transformer version I used is 4.31.0, batch size 8 with accumaltion of 4, lr = 5e-5 with 1500 steps of warmup. It seems that maybe lower lr could be tried first?
Also, do you think the problem be the limited data?

**Reply by luvwinnie** [reply]
I'm having same problem with small Japanese dataset, what parameters should we be careful? Anyone solved the hallucination?

**Comment by toanhuynhnguyen** [comment]
Based on this guide https://huggingface.co/blog/fine-tune-whisper, I tried to fine-tune "small" and "large-v3" models.
- The fine-tuned "small" model works normally.
- But the fine-tuned "large-v3" model works poorly on non-English audio files such as Chinese audio files, it auto-translates Chinese to English though I specified transcribing in Chinese, not translating. Have you faced this issue, and can you give me advice? Thank you so much.
